File: 2017/Matt/Java/src/html/advent2017/Day09.java
package html.advent2017;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

class Day09 extends DayBase {
  private String dataFile = "data/day09";

  void solution() {
    String data = getData();
    System.out.println(part1(data));
    System.out.println(part2(data));
  }


  private String getData() {
    String line = "";

    try {
      FileReader fRead = new FileReader(dataFile);
      BufferedReader bRead = new BufferedReader(fRead);
      line = bRead.readLine();
      bRead.close();
    } catch (IOException e) {
      e.printStackTrace();
    }

    return line;
  }


  private int part1(String data) {
    int score = 0;
    int index = 0;
    int group = 0;
    boolean garbage = false;
    boolean cancel = false;

    do {  // while (group > 0)
      char next = data.charAt(index);
      if (cancel) {
        cancel = false;
      }else {
        if (garbage) {
          if (next == '!') {
            cancel = true;
          } else if (next == '>') {
            garbage = false;
          }
        } else {
          if (next == '{') {
            group++;
            score += group;
          } else if (next == '}') {
            group--;
          } else if (next == '<') {
            garbage = true;
          }
        }
      }
      index++;
    } while (group > 0);
    return score;
  }


  private int part2(String data) {
    int trashCount = 0;
    int index = 0;
    int group = 0;
    boolean garbage = false;
    boolean cancel = false;

    do {  // while (group > 0)
      char next = data.charAt(index);
      if (cancel) {
        cancel = false;
      }else {
        if (garbage) {
          if (next == '!') {
            cancel = true;
          } else if (next == '>') {
            garbage = false;
          } else {
            trashCount++;
          }
        } else {
          if (next == '{') {
            group++;
          } else if (next == '}') {
            group--;
          } else if (next == '<') {
            garbage = true;
          }
        }
      }
      index++;
    } while (group > 0);
    return trashCount;
  }
}

File: 2016/Matt/Kotlin/src/Day09.kt
package html.advent2016

import java.io.File

class Day09 : DayBase() {
    private val dataFile = "data/day09"

    override fun solution() {
        val data = getData()
        println(part1(data))
        println(part2(data))
    }


    private fun getData(): String {
        return File(dataFile).readText()
    }


    private fun part1(data: String): Int {
        var decompressedLength = 0
        var index = 0

        while (index < data.length) {
            if (data[index] == '(') {
                var marker = ""
                index++

                while (data[index] != ')') {
                    marker += data[index]
                    index++
                }

                val repeatLength = marker.split("x")[0].toInt()
                val repeatCount = marker.split("x")[1].toInt()

                decompressedLength += repeatLength * repeatCount
                index += repeatLength

            } else {
                decompressedLength++
            }
            index++
        }

        return decompressedLength
    }


    private fun part2(data: String): Long {
        return decompressedSegmentLength(1, data)
    }


    private fun decompressedSegmentLength(numberOfRepeats: Int, compressedSegment: String): Long {
        var decompressedLength: Long = 0
        var index = 0

        while (index < compressedSegment.length) {
            if (compressedSegment[index] == '(') {
                var marker = ""
                index++

                while (compressedSegment[index] != ')') {
                    marker += compressedSegment[index]
                    index++
                }
                index++

                val repeatLength = marker.split("x")[0].toInt()
                val repeatCount = marker.split("x")[1].toInt()

                decompressedLength += decompressedSegmentLength(repeatCount
                        , compressedSegment.substring(index, index + repeatLength))

                index += repeatLength
            } else {
                decompressedLength++
                index++
            }
        }

        return decompressedLength * numberOfRepeats
    }
}


fun main(args: Array<String>) {
    val day = Day09()
    day.solution()
}